README
NAME
    autovivification - Lexically disable autovivification.

VERSION
    Version 0.15

SYNOPSIS
        no autovivification;

        my $hashref;

        my $a = $hashref->{key_a};       # $hashref stays undef

        if (exists $hashref->{option}) { # Still undef
         ...
        }

        delete $hashref->{old};          # Still undef again

        $hashref->{new} = $value;        # Vivifies to { new => $value }

DESCRIPTION
    When an undefined variable is dereferenced, it gets silently upgraded to
    an array or hash reference (depending of the type of the dereferencing).
    This behaviour is called *autovivification* and usually does what you
    mean (e.g. when you store a value) but it may be unnatural or surprising
    because your variables gets populated behind your back. This is
    especially true when several levels of dereferencing are involved, in
    which case all levels are vivified up to the last, or when it happens in
    intuitively read-only constructs like "exists".

    This pragma lets you disable autovivification for some constructs and
    optionally throws a warning or an error when it would have happened.

METHODS
  "unimport"
        no autovivification; # defaults to qw<fetch exists delete>
        no autovivification qw<fetch store exists delete>;
        no autovivification 'warn';
        no autovivification 'strict';

    Magically called when "no autovivification @opts" is encountered.
    Enables the features given in @opts, which can be :

    *   'fetch'

        Turns off autovivification for rvalue dereferencing expressions,
        such as :

            $value = $arrayref->[$idx]
            $value = $hashref->{$key}
            keys %$hashref
            values %$hashref

        Starting from perl 5.11, it also covers "keys" and "values" on array
        references :

            keys @$arrayref
            values @$arrayref

        When the expression would have autovivified, "undef" is returned for
        a plain fetch, while "keys" and "values" return 0 in scalar context
        and the empty list in list context.

    *   'exists'

        Turns off autovivification for dereferencing expressions that are
        parts of an "exists", such as :

            exists $arrayref->[$idx]
            exists $hashref->{$key}

        '' is returned when the expression would have autovivified.

    *   'delete'

        Turns off autovivification for dereferencing expressions that are
        parts of a "delete", such as :

            delete $arrayref->[$idx]
            delete $hashref->{$key}

        "undef" is returned when the expression would have autovivified.

    *   'store'

        Turns off autovivification for lvalue dereferencing expressions,
        such as :

            $arrayref->[$idx] = $value
            $hashref->{$key} = $value
            for ($arrayref->[$idx]) { ... }
            for ($hashref->{$key}) { ... }
            function($arrayref->[$idx])
            function($hashref->{$key})

        An exception is thrown if vivification is needed to store the value,
        which means that effectively you can only assign to levels that are
        already defined. In the example, this would require $arrayref (resp.
        $hashref) to already be an array (resp. hash) reference.

    *   'warn'

        Emits a warning when an autovivification is avoided.

    *   'strict'

        Throws an exception when an autovivification is avoided.

    Each call to "unimport" adds the specified features to the ones already
    in use in the current lexical scope.

    When @opts is empty, it defaults to "qw<fetch exists delete>".

  "import"
        use autovivification; # default Perl behaviour
        use autovivification qw<fetch store exists delete>;

    Magically called when "use autovivification @opts" is encountered.
    Disables the features given in @opts, which can be the same as for
    "unimport".

    Each call to "import" removes the specified features to the ones already
    in use in the current lexical scope.

    When @opts is empty, it defaults to restoring the original Perl
    autovivification behaviour.

CONSTANTS
  "A_THREADSAFE"
    True if and only if the module could have been built with thread-safety
    features enabled. This constant only has a meaning when your perl is
    threaded, otherwise it will always be false.

  "A_FORKSAFE"
    True if and only if this module could have been built with fork-safety
    features enabled. This constant will always be true, except on Windows
    where it is false for perl 5.10.0 and below.

CAVEATS
    Using this pragma will cause a slight global slowdown of any subsequent
    compilation phase that happens anywere in your code - even outside of
    the scope of use of "no autovivification" - which may become noticeable
    if you rely heavily on numerous calls to "eval STRING".

    The pragma doesn't apply when one dereferences the returned value of an
    array or hash slice, as in "@array[$id]->{member}" or
    @hash{$key}->{member}. This syntax is valid Perl, yet it is discouraged
    as the slice is here useless since the dereferencing enforces scalar
    context. If warnings are turned on, Perl will complain about one-element
    slices.

    Autovivifications that happen in code "eval"'d during the global
    destruction phase of a spawned thread or pseudo-fork (the processes used
    internally for the "fork" emulation on Windows) are not reported.

DEPENDENCIES
    perl 5.8.3.

    A C compiler. This module may happen to build with a C++ compiler as
    well, but don't rely on it, as no guarantee is made in this regard.

    XSLoader (standard since perl 5.6.0).

SEE ALSO
    perlref.

AUTHOR
    Vincent Pit, "<perl at profvince.com>", <http://www.profvince.com>.

    You can contact me by mail or on "irc.perl.org" (vincent).

BUGS
    Please report any bugs or feature requests to "bug-autovivification at
    rt.cpan.org", or through the web interface at
    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=autovivification>. I
    will be notified, and then you'll automatically be notified of progress
    on your bug as I make changes.

SUPPORT
    You can find documentation for this module with the perldoc command.

        perldoc autovivification

    Tests code coverage report is available at
    <http://www.profvince.com/perl/cover/autovivification>.

ACKNOWLEDGEMENTS
    Matt S. Trout asked for it.

COPYRIGHT & LICENSE
    Copyright 2009,2010,2011,2012,2013,2014,2015 Vincent Pit, all rights
    reserved.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for autovivification

0.15    2015-05-14 20:35 UTC
        + Chg : The new environment variable to enable thread tests on older
                perls is PERL_FORCE_TEST_THREADS. Note that this variable
                should only be turned on by authors.
        + Fix : The module has been taught about the new multideref
                optimization, and will function properly on perl 5.21.11 and
                above.
        + Fix : The module will no longer crash or behave erratically when it
                is loaded by several threads (or Windows emulated processes)
                ran in parallel.
        + Fix : Update the Windows ActivePerl + gcc 3.4 workaround for
                ExtUtils::MakeMaker 7.04. Thanks Christian Walde for reporting
                and feedback on this issue.
        + Fix : Be really compatible with the optional OP_PARENT feature.
        + Tst : $ENV{$Config{ldlibpthname}} is now preserved on all platforms,
                which will address failures of t/43-peep.t and
                t/51-threads-teardown.t with unusual compilers (like icc) that
                link all their compiled objects to their own libraries.

0.14    2014-11-01 22:25 UTC
        + Fix : [RT #99458] : AnyEvent::Loop hangs if autovivification 0.13 is
                loaded
                [RT #99904] : 'no autovivification' breaks Lingua::EN::Inflect
                from a distance
                This was a regression in 0.13 which caused the peephole
                optimizer to loop infinitely on "... while 1" constructs.
                However, the Lingua::EN::Inflect problem only occured in perl
                5.20 and above.
                Thanks Toshio Ito and Kent Fredric for reporting.

0.13    2014-10-04 16:55 UTC
        This release contains a change that, while being very likely to be safe,
        can potentially cause freezes during code compilation. Every release
        should be carefully tested before being put in production, but this is
        especially true for this one.
        + Add : Support for the PERL_OP_PARENT optional feature introduced in
                perl 5.21.2.
        + Doc : The CAVEATS section now warns about the global slowdown during
                compilation caused by this pragma.
        + Fix : [RT #97703] : Android support
                t/51-threads-teardown.t will no longer fail on Android.
                Thanks Brian Fraser for reporting.
        + Fix : Segfaults in eval in an END block of a Win32 pseudo-fork.
        + Fix : Segfaults during global destruction of a thread or a
                pseudo-fork.
        + Opt : The global slowdown caused by this module has been greatly
                reduced.
                Thanks Ævar Arnfjörð Bjarmason for reporting and testing the
                change.

0.12    2013-09-05 17:20 UTC
        + Fix : Check functions are now replaced and restored in a thread-safe
                manner, either by using the wrap_op_checker() function from perl
                when it is available (starting from perl 5.16) or by taking the
                OP_REFCNT mutex on older perls.
        + Tst : Author tests are no longer bundled with this distribution.
                They are only made available to authors in the git repository.

0.11    2013-02-08 19:25 UTC
        This is a maintenance release. The code contains no functional change.
        Satisfied users of version 0.10 can skip this update.
        + Doc : POD tweaks and clarifications.
        + Tst : Threads tests will not fail anymore if resources constraints
                prevent the system from creating all the required threads.
        + Tst : Author tests overhaul.

0.10    2011-08-24 15:00 UTC
        + Fix : The pragma no longer vivifies the "autovivification" entry
                in the hints hash %^H on perl 5.8.

0.09    2011-01-05 18:40 UTC
        + Fix : [RT #64435] : Hangs with File::Copy in Config.pm.
                This was actually a regression introduced together with the new
                peephole optimizer strategy, and that caused the pragma to hang
                on constructs like "for (;;) { ... }".
                Thanks Michael Schilli for reporting.

0.08    2011-01-03 21:00 UTC
        + Fix : Building on Windows.

0.07    2010-12-31 16:20 UTC
        + Chg : perl 5.8.3 is required.
        + Doc : Complements and clarifications.
        + Fix : Segmentation faults and misbehaviours in threaded applications.
        + Fix : Compatibility with perl 5.13.7.
                Thanks Andreas J. König for reporting and Andrew Main for
                providing a fix.
        + Fix : Broken linkage on Windows with gcc 3.4, which appears in
                particular when using ActivePerl's default compiler suite.
                For those setups, the autovivification shared library will now
                be linked against the perl dll directly (instead of the import
                library).
        + Opt : The pragma takes slightly more time at compile-time, but is
                slightly faster at run-time.
        + Tst : Lengthy tests have been ported to Test::Leaner, making the
                whole test suite about 50% faster.
        + Tst : Threads tests are now only run on perl 5.13.4 and higher.
                They could segfault randomly because of what seems to be an
                internal bug of Perl, which has been addressed in 5.13.4.
                There is also an environment variable that allows you to
                forcefully run those tests, but it should be set only for
                author testing and not for end users.

0.06    2010-04-24 17:40 UTC
        + Add : The A_THREADSAFE and A_FORKSAFE constants.
        + Fix : [RT #56870] : "no autovivification" vs Regexp::Common.
                This was a bug in how tied arrays and hashes were handled.
                Thanks Michael G. Schwern for reporting.

lib/autovivification.pm
package autovivification;

use 5.008_003;

use strict;
use warnings;

=head1 NAME

autovivification - Lexically disable autovivification.

=head1 VERSION

Version 0.15

=cut

our $VERSION;
BEGIN {
 $VERSION = '0.15';
}

=head1 SYNOPSIS

    no autovivification;

    my $hashref;

    my $a = $hashref->{key_a};       # $hashref stays undef

    if (exists $hashref->{option}) { # Still undef
     ...
    }

    delete $hashref->{old};          # Still undef again

    $hashref->{new} = $value;        # Vivifies to { new => $value }

=head1 DESCRIPTION

When an undefined variable is dereferenced, it gets silently upgraded to an array or hash reference (depending of the type of the dereferencing).
This behaviour is called I<autovivification> and usually does what you mean (e.g. when you store a value) but it may be unnatural or surprising because your variables gets populated behind your back.
This is especially true when several levels of dereferencing are involved, in which case all levels are vivified up to the last, or when it happens in intuitively read-only constructs like C<exists>.

This pragma lets you disable autovivification for some constructs and optionally throws a warning or an error when it would have happened.

=cut

BEGIN {
 require XSLoader;
 XSLoader::load(__PACKAGE__, $VERSION);
}

=head1 METHODS

=head2 C<unimport>

    no autovivification; # defaults to qw<fetch exists delete>
    no autovivification qw<fetch store exists delete>;
    no autovivification 'warn';
    no autovivification 'strict';

Magically called when C<no autovivification @opts> is encountered.
Enables the features given in C<@opts>, which can be :

=over 4

=item *

C<'fetch'>

Turns off autovivification for rvalue dereferencing expressions, such as :

    $value = $arrayref->[$idx]
    $value = $hashref->{$key}
    keys %$hashref
    values %$hashref

Starting from perl C<5.11>, it also covers C<keys> and C<values> on array references :

    keys @$arrayref
    values @$arrayref

When the expression would have autovivified, C<undef> is returned for a plain fetch, while C<keys> and C<values> return C<0> in scalar context and the empty list in list context.

=item *

C<'exists'>

Turns off autovivification for dereferencing expressions that are parts of an C<exists>, such as :

    exists $arrayref->[$idx]
    exists $hashref->{$key}

C<''> is returned when the expression would have autovivified.

=item *

C<'delete'>

Turns off autovivification for dereferencing expressions that are parts of a C<delete>, such as :

    delete $arrayref->[$idx]
    delete $hashref->{$key}

C<undef> is returned when the expression would have autovivified.

=item *

C<'store'>

Turns off autovivification for lvalue dereferencing expressions, such as :

    $arrayref->[$idx] = $value
    $hashref->{$key} = $value
    for ($arrayref->[$idx]) { ... }
    for ($hashref->{$key}) { ... }
    function($arrayref->[$idx])
    function($hashref->{$key})

An exception is thrown if vivification is needed to store the value, which means that effectively you can only assign to levels that are already defined.
In the example, this would require C<$arrayref> (resp. C<$hashref>) to already be an array (resp. hash) reference.

=item *

C<'warn'>

Emits a warning when an autovivification is avoided.

=item *

C<'strict'>

Throws an exception when an autovivification is avoided.

=back

Each call to C<unimport> B<adds> the specified features to the ones already in use in the current lexical scope.

When C<@opts> is empty, it defaults to C<< qw<fetch exists delete> >>.

=cut

my %bits = (
 strict => A_HINT_STRICT,
 warn   => A_HINT_WARN,
 fetch  => A_HINT_FETCH,
 store  => A_HINT_STORE,
 exists => A_HINT_EXISTS,
 delete => A_HINT_DELETE,
);

sub unimport {
 shift;
 my $hint = _detag($^H{+(__PACKAGE__)}) || 0;
 @_ = qw<fetch exists delete> unless @_;
 $hint |= $bits{$_} for grep exists $bits{$_}, @_;
 $^H |= 0x00020000;
 $^H{+(__PACKAGE__)} = _tag($hint);
 ();
}

=head2 C<import>

    use autovivification; # default Perl behaviour
    use autovivification qw<fetch store exists delete>;

Magically called when C<use autovivification @opts> is encountered.
Disables the features given in C<@opts>, which can be the same as for L</unimport>.

Each call to C<import> B<removes> the specified features to the ones already in use in the current lexical scope.

When C<@opts> is empty, it defaults to restoring the original Perl autovivification behaviour.

=cut

sub import {
 shift;
 my $hint = 0;
 if (@_) {
  $hint = _detag($^H{+(__PACKAGE__)}) || 0;
  $hint &= ~$bits{$_} for grep exists $bits{$_}, @_;
 }
 $^H |= 0x00020000;
 $^H{+(__PACKAGE__)} = _tag($hint);
 ();
}

=head1 CONSTANTS

=head2 C<A_THREADSAFE>

True if and only if the module could have been built with thread-safety features enabled.
This constant only has a meaning when your perl is threaded, otherwise it will always be false.

=head2 C<A_FORKSAFE>

True if and only if this module could have been built with fork-safety features enabled.
This constant will always be true, except on Windows where it is false for perl 5.10.0 and below.

=head1 CAVEATS

Using this pragma will cause a slight global slowdown of any subsequent compilation phase that happens anywere in your code - even outside of the scope of use of C<no autovivification> - which may become noticeable if you rely heavily on numerous calls to C<eval STRING>.

The pragma doesn't apply when one dereferences the returned value of an array or hash slice, as in C<< @array[$id]->{member} >> or C<< @hash{$key}->{member} >>.
This syntax is valid Perl, yet it is discouraged as the slice is here useless since the dereferencing enforces scalar context.
If warnings are turned on, Perl will complain about one-element slices.

Autovivifications that happen in code C<eval>'d during the global destruction phase of a spawned thread or pseudo-fork (the processes used internally for the C<fork> emulation on Windows) are not reported.

=head1 DEPENDENCIES

L<perl> 5.8.3.

A C compiler.
This module may happen to build with a C++ compiler as well, but don't rely on it, as no guarantee is made in this regard.

L<XSLoader> (standard since perl 5.6.0).

=head1 SEE ALSO

L<perlref>.

=head1 AUTHOR

Vincent Pit, C<< <perl at profvince.com> >>, L<http://www.profvince.com>.

You can contact me by mail or on C<irc.perl.org> (vincent).

=head1 BUGS

Please report any bugs or feature requests to C<bug-autovivification at rt.cpan.org>, or through the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=autovivification>.
I will be notified, and then you'll automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc autovivification

Tests code coverage report is available at L<http://www.profvince.com/perl/cover/autovivification>.

=head1 ACKNOWLEDGEMENTS

Matt S. Trout asked for it.

=head1 COPYRIGHT & LICENSE

Copyright 2009,2010,2011,2012,2013,2014,2015 Vincent Pit, all rights reserved.

This program is free software; you can redistribute it and/or modify it under the same terms as Perl itself.

=cut

1; # End of autovivification
